### ISSUE TITLE
Design a many-to-many relation for Likes 

### ISSUE THREAD

**Issue by hudson221b** [hudson221b/nextmatch#4]
## Business request
An important feature of a dating app is to allow its members to like other members and keep track of whom they have liked and liked by others. 

### Pattern A:
From a member's point of view we think: members I like are certainly tied with "me", so are members that like me. To express such relation, we can add two new fields to the `Member` model:
```
model Member {
...
sourceLike: [memberId1, memberId2,....] //members I like,  I'm the source of likes
targetLike: [memberId1, memberId2,....] // members that like me, I'm the target of likes
}
```
### Pattern B:
We can create a new model to represent the 'liking' behavior:
```
model Like {
  sourceMemberId String 
  sourceMember Member @relation("source", fields:[sourceMemberId], references: [memberId], onDelete: Cascade)

  targetMemberId String
  targetMember Member @relation("target", fields:[targetMemberId], references: [memberId], onDelete: Cascade)
}
```
And associate Like with members like this:
```
model Member {
  sourceLike  Like[] @relation("source")
  targetLike  Like[] @relation("target")
}
```

## Considerations and decision
Pros of pattern A:
- simplicity: all information is contained within a single model.
- quick access: you can easily retrieve a member's likes without additional queries.
Cons of pattern A:
- limited scalability: as the number of likes grows, the Member document size increases, potentially hitting document size limits
- slow updates: adding or removing likes requires updating the entire array
- lack of additional metadata: It's harder to add extra information about the like (e.g., timestamp, status).
- redundancy: The same information is stored in two places (A like B, B's ID is stored in `sourceIds` on A, and A's ID is stored in `targetIds` on B)

Pattern B overcomes the cons of pattern A in industrial-level products in that:
- it handles a large number of Likes without affecting the Member model
- it follows the "atom principle" much better. In pattern A, the liking behavior needs to be stored in two places. What if the update to A's record succeeded but update to B's record failed? It could easily lead to data inconsistencies. In pattern B, each Like is represented by a single record. This not only reduces redundancy but also reducing the risk of data inconsistency. 

**In conclusion, pattern B aligns better with database normalization principles and provides more flexibility despite potential more queries or added complexity to maintain**
